src/varstore.rs

```diff
@@ -0,0 +1,359 @@
+use std::collections::BTreeMap;
+
+use crate::attribute::EfiAttribute;
+use crate::status::EfiStatus;
+
+#[derive(Debug, Clone)]
+pub struct EfiVariable {
+    pub data: Vec<u8>,
+    pub attr: EfiAttribute,
+}
+
+impl EfiVariable {
+    fn new(data: Vec<u8>, attr: EfiAttribute) -> EfiVariable {
+        EfiVariable { data, attr }
+    }
+}
+
+#[derive(Debug)]
+pub struct Varstore {
+    variables: BTreeMap<String, EfiVariable>,
+    bootservices_exited: bool,
+    max_data_length: usize,
+    max_name_length: usize,
+}
+
+impl Varstore {
+    pub fn new() -> Varstore {
+        Varstore {
+            variables: BTreeMap::new(),
+            bootservices_exited: false,
+            max_name_length: 10000,
+            max_data_length: 50000,
+        }
+    }
+
+    pub fn with_limits(max_name: usize, max_data: usize) -> Varstore {
+        Varstore {
+            variables: BTreeMap::new(),
+            bootservices_exited: false,
+            max_data_length: max_data,
+            max_name_length: max_name,
+        }
+    }
+
+    fn request_get_next(&self, name: Option<String>) -> Option<(&String, &EfiVariable)> {
+        match name {
+            Some(var_name) => self
+                .variables
+                .iter()
+                .skip_while(|&(k, _v)| k != &var_name)
+                .nth(1),
+            None => self.variables.iter().next(),
+        }
+    }
+
+    fn request_get(&self, name: &String) -> Option<&EfiVariable> {
```

Comment:
Similar to the comment on `request_get_next`, this could take name as `&str` because most stringy types can be freely cast into `&str` but not into `&String`.